src/physics/initial_conditions/set_initial_condition.cpp

```diff
@@ -83,6 +101,144 @@ void SetInitialCondition<dim,nstate,real>::project_initial_condition(
+    
+    // (1) Get filename based on MPI rank
+    //-------------------------------------------------------------
+    const int mpi_rank = dealii::Utilities::MPI::this_mpi_process(MPI_COMM_WORLD);
+    // -- Get padded mpi rank string
+    const std::string mpi_rank_string = get_padded_mpi_rank_string(mpi_rank);
+    // -- Assemble filename string
+    const std::string filename_without_extension = input_filename_prefix + std::string("-") + mpi_rank_string;
+    const std::string filename = filename_without_extension + std::string(".dat");
+    //-------------------------------------------------------------
+
+    // (2) Read file
+    //-------------------------------------------------------------
+    std::string line;
+    std::string::size_type sz1;
+
+    std::ifstream FILE (filename);
+    std::getline(FILE, line); // read first line: DOFs
+    
+    // check that the file is not empty
+    if (line.empty()) {
+        pcout << "ERROR: Trying to read empty file named " << filename << std::endl;
+        std::abort();
+    } else {
+        const unsigned int number_of_degrees_of_freedom_per_state_DG = dg->dof_handler.n_dofs()/nstate;
+        const unsigned int number_of_degrees_of_freedom_per_state_file = std::stoi(line);
+        if(number_of_degrees_of_freedom_per_state_file != number_of_degrees_of_freedom_per_state_DG) {
+            pcout << "ERROR: Cannot read initial condition. "
+                      << "Number of degrees of freedom per state do not match expected by DG in file: " 
+                      << filename << "\n Aborting..." << std::endl;
+            std::abort();
+        }
+    }
+
+    std::getline(FILE, line); // read first line of data
+
+    // check that there indeed is data to be read
+    if (line.empty()) {
+        pcout << "Error: File has no data to be read" << std::endl;
+        std::abort();
+    }
+
+    // Commented since this has not yet been tested
+    // dealii::LinearAlgebra::distributed::Vector<double> solution_no_ghost;
+    // solution_no_ghost.reinit(dg->locally_owned_dofs, MPI_COMM_WORLD);
+    // dealii::AffineConstraints affine_constraints(dof_handler.locally_owned_dofs());
+    // dealii::VectorTools::project(*(dg->high_order_grid->mapping_fe_field),dg->dof_handler,affine_constraints,dg->volume_quadrature_collection,*initial_condition_function,solution_no_ghost);
+    // dg->solution = solution_no_ghost;
+
+    //Note that for curvilinear, can't use dealii interpolate since it doesn't project at the correct order.
+    //Thus we interpolate it directly.
+    const auto mapping = (*(dg->high_order_grid->mapping_fe_field));
+    dealii::hp::MappingCollection<dim> mapping_collection(mapping);
+    dealii::hp::FEValues<dim,dim> fe_values_collection(mapping_collection, dg->fe_collection, dg->volume_quadrature_collection, 
+                                dealii::update_quadrature_points);
+    const unsigned int max_dofs_per_cell = dg->dof_handler.get_fe_collection().max_dofs_per_cell();
```

Comment:
Does this projection work for meshes with non-constant polynomial order?

Reply:
This function is a copy of `project_initial_condition` function but with the file reading stuff inside it since that is done on the fly -- I could merge the two functions into one but then there would be `if` statements all over the place; up to you what you think is best

Reply:
I think it's better to leave this as-is in terms of keeping it separate from `project_initial_condition`.

What I meant is to do with the mesh. If some cells have different `p`, can we still use the same volume projection operator? Or should we check `dg->get_max_fe_degree() == dg->get_min_fe_degree()`? I'm not familiar with the operators. (If you feel this case is not common enough to check for, no problem to leave the function as-is).

Reply:
Ahh I see, yeah I'm not sure

Reply:
Per discussion, we will keep this in mind as a potential issue if we need to set initial conditions on a grid that has non-uniform `p`.